tests/test_fit_result.py

```diff
@@ -15,89 +17,92 @@
     LogLikelihood, LeastSquares, VectorLeastSquares, MinimizeModel
 )
 
-@pytest.fixture
-def a():
-    return Parameter('a')
-
 
-@pytest.fixture
-def b():
-    return Parameter('b')
+# TODO Entfernen
+# 1. Likelihood & Constraints Ergebnisse (a, b, (r²)) verschieden >>> mit wolfram alpha testen?
+# 2. kein r² bei likelihood >>> check mit hasttr
+# 3. Chained Minimizer & LeastSquare >>> einfach einsetzen, obwohl es nicht der normalfall ist?
+# 4. r² zu hasttr Sektion hinzufügen? >>> why not
+# 5. Zukunft: test_fitting_2 simpler/zu großem Test?    
 
+def ge_constraint(a):  # Has to be in the global namespace for pickle.
+    return a - 1
 
-@pytest.fixture
-def result_dict(a, b):
-
-    result_dict = dict()
 
-    z = Variable('z')
-    x = Variable('x')
-    y = Variable('y')
-    xdata = np.linspace(1, 10, 10)
-    ydata = 3 * xdata ** 2
+constraints = [
+    Eq(a, b),
+    CallableNumericalModel.as_constraint(
+        {z: ge_constraint}, connectivity_mapping={z: {a}},
+        constraint_type=Ge, model=Model({y: a * x ** b})
+    )
+]
+
+
+@pytest.mark.parametrize('kwargs, objective_name, gof_presence', [
+    ({'x': np.linspace(1, 10, 10), 'y': 3 * np.linspace(1, 10, 10) ** 2},
+     LeastSquares, (True, True, True, False, False)),
+    ({'x': np.linspace(1, 10, 10), 'y': 3 * np.linspace(1, 10, 10) ** 2,
+      'minimizer': MINPACK}, VectorLeastSquares, (True, True, True, False, False)),
+    ({'x': np.linspace(1, 10, 10), 'objective': LogLikelihood},
+     LogLikelihood, (True, False, False, True, True)),
+    ({'x': np.linspace(1, 10, 10), 'y': 3 * np.linspace(1, 10, 10) ** 2, 'minimizer': [BFGS, NelderMead]},
+     LeastSquares, (True, True, True, False, False)),
+    ({'x': np.linspace(1, 10, 10), 'y': 3 * np.linspace(1, 10, 10) ** 2,
+      'constraints': constraints}, LeastSquares, (True, True, True, False, False)),
+    ({'x': np.linspace(1, 10, 10), 'y': 3 * np.linspace(1, 10, 10) ** 2, 'constraints': constraints, 'minimizer': BasinHopping},
+     LeastSquares, (True, True, True, False, False))
```

Comment:
Let's see if we can dissect what's going on in these tests.

1) 2 different objectives: LeastSquares and LogLikelihood. If LS is used both an x and a y are passed, with LL only x is passed.
2) A few different minimizers are used
3) Sometimes constraints are used, but not all minimizers can deal with those.
4) The resulting FitResults have different attributes/values in the end.

Although 1 and 2+3 are orthogonal (i.e. could be separate `parametrize` calls), but 4 will be different for all combinations.

Idea 1: make a big table:
```python
@pytest.mark.parametrize('fit_kwargs, param_values, param_std, gof_values', [
    (
       {'x': ..., 'y': ..., 'objective': LeastSquares, 'minimizer': ..., 'constraints': ...}
       {'a': ..., 'b': ...}, {'a': ..., 'b': ...}
       {'r_squared': 0.99, ...}
    ),
    .
    .
    .
])
def test_everything(...):
```
Pro: minimal number of tests, easy
Con: Maximum number of parametrizations, missing combinations.

Idea 2: Separate out the special cases: LogLikelihood, and constraints.
```python
MINIMIZERS = subclasses(BaseMinimizer)  # https://github.com/tBuLi/symfit/blob/ecc567a1e75092dc3c94cecb0b44c132dbab7f4e/tests/test_minimizers.py#L63
CONSTR_MINIMIZERS = subclasses(ConstrainedMinimizer)

@pytest.mark.parametrize('minimizer', 
    MINIMIZERS
)
def test_least_sq_no_constraint(minimizer):
    objective = LeastSquares
    expected = {'a': 3, 'b': 2}
    expected_std = {'a': ..., 'b': ...}
    expected_gof = {'r_squared': 0.99, ...}
    fit_kwargs = dict(objective=objective, minimizer=minimizer, x=..., y=...)
    _run_test(fit_kwargs, expected, expected_std, expected_gof)

@pytest.mark.parametrize('minimizer', 
    CONSTR_MINIMIZERS
)
def test_least_sq_constraint(minimizer):
    objective = LeastSquares
    expected = {'a': ..., 'b': ...}
    expected_std = {'a': ..., 'b': ...}
    expected_gof = {'r_squared': 0.99, ...}
    fit_kwargs = dict(objective=objective, minimizer=minimizer, x=..., y=..., constraints=...)
    _run_test(fit_kwargs, expected, expected_std, expected_gof)

@pytest.mark.parametrize('minimizer', 
    MINIMIZERS
)
def test_likelihood_no_constr(minimizer):
    objective = LogLikelihood
    expected = {'a': ..., 'b': ...}
    expected_std = {'a': ..., 'b': ...}
    expected_gof = {...}
    fit_kwargs = dict(objective=objective, minimizer=minimizer, x=..., y=..., constraints=...)
    _run_test(fit_kwargs, expected, expected_std, expected_gof)

# etc, etc

def _run_test(fit_kwargs, expected, expected_std, expected_gof):
    fit = Fit(**fit_kwargs)
    ... # assert all the things!
```
CON: many different tests, some code duplication
PRO: probably best coverage of all different cases, minimal duplication of data/cases


At the moment I think idea 2 has my preference. What do you think?

Reply:
I just had a few minutes. But I want to share my first impression: I do like the second one too! But it sounds like we transfer the parameterization into methods, right? On the other hand, it's a very pretty workaround for nested parameterization 😃

Reply:
> But it sounds like we transfer the parameterization into methods, right?

What do you mean? I just made 2 variables in the global namespace that contain lists of minimizers.

Reply:
Just for clearance: At the position of "`# etc etc`" would be `test_likelihood_constraints`, the special kid `ChainedMinimizer` (or would it already be included in the LL tests?) and tests without any specifications of minimizer or objective, but with and without constraints. In every case, we have to pass and check expected minimizers and objectives...right?
As a result, we would have *many* tests, which are not "pytest-style", but the coverage would be 👍 👍 👍 

PS: The [test](https://github.com/DuckNrOne/symfit/blob/f8bd87278238f79c533e516d1e6bd31e60723521/tests/test_fit_result.py#L76) that checks, whether the minimizer is an instance of `BaseMinimizer`, could be specified with the check for `ConstrainedMinimizer`, right?

Reply:
> What do you mean? I just made 2 variables in the global namespace that contain lists of minimizers.

You declare variables (fit_kwargs, expected, expected_std, expected_gof) and call a method (_run_test). In the first moment, it sounds like parameterization. We could combine the tests with `BaseMinimizer` to a big test
```python
MINIMIZERS = subclasses(BaseMinimizer)

@pytest.mark.parametrize('minimizer', 
    MINIMIZERS
)
@pytest.mark.parametrize('fit_kwargs, expected, expected_std, expected_got',
    [
    (dict(objective=LeastSquares, x=..., y=...), {'a': 3, 'b': 2}, {'a': ..., 'b': ...}, {'r_squared': 0.99, ...}),
    (dict(objective=LogLikelihood, x=...),  {'a': ..., 'b': ...}, {'a': ..., 'b': ...}, {...})
])
def test_no_constraint(minimizer, fit_kwargs, expected, expected_std, expected_got):
    # But we still have to call another method (like _run_tests) to minimize duplicated code. Like:
    fit = Fit(**fit_kwargs, minimizer=minimizer)
    _run_tests(fit, expected, expected_std, expected_got)
```
We could combine tests with `ConstraintMinimizer`. BUT we would lose legibility and transparency in comparison to your second idea...

Reply:
> Just for clearance: At the position of "# etc etc" would be test_likelihood_constraints, the special kid ChainedMinimizer...

Yes

> ... tests without any specifications of minimizer or objective, but with and without constraints. 

Add None to the minimizer lists

> In every case, we have to pass and check expected minimizers and objectives...right?

I guess so. This was a rough sketch. There's probably a few things I missed.

> As a result, we would have many tests, which are not "pytest-style

Still fewer than there are now. Especially since the actual test would still be one function.

> The test that checks, whether the minimizer is an instance of BaseMinimizer, could be specified with the check for ConstrainedMinimizer, right?

If you have access to the minimizer given you can tighten the test by checking the minimizer passed as argument is used/propagated. But every (Constrained)Minimizer is a BaseMinimizer.

> You declare variables (fit_kwargs, expected, expected_std, expected_gof) and call a method (_run_test). In the first moment, it sounds like parameterization. We could combine the tests with BaseMinimizer to a big test

Clever, I like it. Be sure to take enough vertical space to lay-out the arguments nicely and clearly separate the LL from LS.
Actually, if you add a Model to those parameters you suddenly create a *very* thorough test... 🤔 But then you'd also need the initial variable values. Maybe for later. Let's remember this idea.

Reply:
Thank you for the detailed feedback :D 
Conclusion? Which idea should implement?

**Idea 1**
Pro: minimal number of tests, easy, pytest-style
Con: Maximum number of parametrizations, missing combinations.

**Idea 2**
Pro: probably best coverage of all different cases, minimal duplication of data/cases, transparent and readable code
Con: many different tests, some code duplication

**Idea 3**
Pro: probably *best* coverage of all different cases, pytest-style, no code duplication
Con: lose readability and transparency, many different tests

Reply:
> Thank you for the detailed feedback :D

I have a vested interest in making these tests work (:

I'd say number 3. But if you halfway through find that it's not quite convenient feel free to go for 2. Or 4.

Reply:
Funny story: probl...obstacles :D

1. I don't know if the method `subclasses` is the best solution. It includes `ChainedMinimizer`, which causes an error because it's initialized without any minimizer. Maybe we have to list them in the Parametrization, like
```python
@pytest.mark.parametrize('minimizer', 
    [MINPACK, BFGS, NelderMead, ChainedMinimizer, BasinHopping, ...]
)
```
2. Such a list would help at the next point too: "Add None to the minimizer lists" sound easier as it is :D Because `subclasses` returns a set, it's difficult to combine it with `None`. With a simple Parametrization, we could add `None` (and a ChainedMinimizer instance) like
```python
@pytest.mark.parametrize('minimizer', 
    [MINPACK, BFGS, NelderMead, ChainedMinimizer, BasinHopping, ..., None, [BFGS, NelderMead]]
)
```
3. Another obstacle we have to overcome is `stdev`. Because we use different minimizers, we have different stdevs. We could overcome it by using `pytest.approx(value, abs=1e-3`) for example. Easy...
4. A major obstacle is the deviating values *a* and *b* with the minimizers `COBYLA`, `TrustConstr`. 
5. DifferentialEvolution makes problems together with scipy: "ValueError: bounds should be a sequence containing real valued (min, max) pairs for each value in x"...Probably because of the model (?)
6. What should we do with our next "problem child": `MINPACK` and `VectorLeastSquares`? Own test, but with `_run_test` in my opinion :D

Reply:
1+2. You can do e.g. `MINIMIZERS = subclasses(...) - {ChainedMinimizer} + {None}`. Sets are mutable.
3. This is very strange. The stdev only depends on the model, data and the found solution. Not on the minimizer.
4. Also very interesting. How much are they off? It could be a bug in how we call the minimizer, could be a bug in the minimizers themselves...
5. DiffEvo requires bounds on all the parameters. Leave it out of these tests I suggest.
6. Agreed.

Reply:
I uploaded a new draft with some improvements:

> 4. How much are they off?

With `objective=LogLikelihood` no result occurs twice. With the other objectives, the result is 2.3 instead of 3.

> 3. The stdev only depends on the model, data, objective, and the found solution.

The uploaded draft includes the solution with `pytest.approx(value, abs=1e-3)`.

> 6. Agreed.

The only difference between MINPACK and the other minimizers is the "default objective". So I would use the same pattern as `test_constraints' and `test_no_constraints', except that there is no parameterization of the minimizers, and VectorLeastSquares is asserted.

btw, I added a ChainedMinimizer as a tuple to this line: `MINIMIZERS = subclasses(...) - {ChainedMinimizer} + {None, ((BFGS, NelderMead)}`. I hope this is legal :D

Reply:
> 4. the result is 2.3 instead of 3
Very strange. I'll try to find some time to investigate this further, but it may take some time.

Reply:
Working on the special test for `MINPACK`: If I assign `objective` to `LogLikelihood` or `LeastSquares` the following error accours:
```
Traceback (most recent call last):
  File "test.py", line 12, in <module>
    fit_result = fit.execute()
  File "/home/antony/.local/lib/python2.7/site-packages/symfit/core/fit.py", line 579, in execute
    minimizer_ans = self.minimizer.execute(**minimize_options)
  File "/home/antony/.local/lib/python2.7/site-packages/symfit/core/minimizers.py", line 788, in execute
    **minpack_options
  File "/home/antony/.local/lib/python2.7/site-packages/symfit/core/leastsqbound.py", line 280, in leastsqbound
    raise TypeError('Improper input: N=%s must not exceed M=%s' % (n, m))
TypeError: Improper input: N=2 must not exceed M=1
```
N? M?

Reply:
You cant use the "normal" LeastSquares, nor LL objectives with MINPACK, since the minimizer needs access to the residuals rather than just the objective value, which is why VectorLeastSquares exists.

Reply:
Ah! That explains allot. 
That means, that we're able to check MINPACK only with VectorLeastSquares. Great :D